src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -113,84 +114,82 @@ public void Start<TContext>(IHttpApplication<TContext> application)
                 engine.Start(threadCount);
                 var atLeastOneListener = false;
 
-                foreach (var address in _serverAddresses.Addresses.ToArray())
-                {
-                    var parsedAddress = ServerAddress.FromUrl(address);
-                    atLeastOneListener = true;
+                var listenOptions = Options.ListenOptions;
 
-                    if (!parsedAddress.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+                if (listenOptions.Any())
+                {
+                    var addresses = _serverAddresses.Addresses;
+                    if (addresses.SingleOrDefault() != "http://localhost:5000")
```

Comment:
weird guesswork that makes assumptions about the intent of the caller. I'd rather reconsider the Hosting default logic. WebListener is going to have a similar issue when it's configured to attach to an existing queue. @davidfowl